## Analyzing data on gun deaths in the US

The dataset came from FiveThirtyEight, and can be found https://github.com/fivethirtyeight/guns-data.  
The dataset is stored in the guns.csv file.  
It contains information on gun deaths in the US from 2012 to 2014. Each row in the dataset represents a single fatality.  
The columns contain demographic and other information about the victim. Here are the first few rows of the dataset:

## A) Read the dataset in as a list using the csv module.
 1. Import the csv module.
 2. Open the file using the open() function.
 3. Use the csv.reader() function to load the opened file.
 4. Call list() on the result to get a list of all the data in the file.
 5. Assign the result to the variable data.
 6.Display the first 5 rows of data to verify everything.

In [1]:
import csv 
f = open('guns.csv',"r")
csv_reader = csv.reader(f)
data = list(csv_reader)

print(data[0:5])

[['', 'year', 'month', 'intent', 'police', 'sex', 'age', 'race', 'hispanic', 'place', 'education'], ['1', '2012', '01', 'Suicide', '0', 'M', '34', 'Asian/Pacific Islander', '100', 'Home', '4'], ['2', '2012', '01', 'Suicide', '0', 'F', '21', 'White', '100', 'Street', '3'], ['3', '2012', '01', 'Suicide', '0', 'M', '60', 'White', '100', 'Other specified', '4'], ['4', '2012', '02', 'Suicide', '0', 'M', '64', 'White', '100', 'Home', '4']]


## B) Removing Headers From A List Of Lists

1. Extract the first row of data, and assign it to the variable headers.
2. Remove the first row from data.
3. Display headers.
4. Display the first 5 rows of data to verify that you removed the header row properly.

In [2]:
headers = data[0]
data = data[1:]

print(headers)
print(data[0:5])


['', 'year', 'month', 'intent', 'police', 'sex', 'age', 'race', 'hispanic', 'place', 'education']
[['1', '2012', '01', 'Suicide', '0', 'M', '34', 'Asian/Pacific Islander', '100', 'Home', '4'], ['2', '2012', '01', 'Suicide', '0', 'F', '21', 'White', '100', 'Street', '3'], ['3', '2012', '01', 'Suicide', '0', 'M', '60', 'White', '100', 'Other specified', '4'], ['4', '2012', '02', 'Suicide', '0', 'M', '64', 'White', '100', 'Home', '4'], ['5', '2012', '02', 'Suicide', '0', 'M', '31', 'White', '100', 'Other specified', '2']]


## C) Index of column and Information contained herein

1. '' - this is an identifier column, which contains the row number. It's common in CSV files to include a unique identifier for each row, but we can ignore it in this analysis.
2. year -- the year in which the fatality occurred.
3. month -- the month in which the fatality occurred.
4. intent -- the intent of the perpetrator of the crime. This can be Suicide, Accidental, NA, Homicide, or Undetermined.
5. police -- whether a police officer was involved with the shooting. Either 0 (false) or 1 (true).
6. sex -- the gender of the victim. Either M or F.
7. age -- the age of the victim. 
8. race -- the race of the victim. Either Asian/Pacific Islander, Native American/Native Alaskan, Black, Hispanic, or White
9. hispanic -- a code indicating the Hispanic origin of the victim
10. place -- where the shooting occurred. Has several categories, which you're encouraged to explore on your own.
11. education -- educational status of the victim. Can be one of the following:
      + 1 -- Less than High School
      + 2 -- Graduated from High School or equivalent
      + 3 -- Some College
      + 4 -- At least graduated from College
      + 5 -- Not available


## D) Counting Gun Deaths By Year

1. Use a list comprehension to extract the year column from data.  
   Because the year column is the second column in the data, you'll need to get the element at index 1 in each row.  
   Assign the result to the variable years.

2. Create an empty dictionary called year_counts.

3. Loop through each element in years.If the element isn't a key in year_counts, create it, and set the value to 1.
   If the element is a key in year_counts, increment the value by one. Display year_counts to see how many gun deaths 
   occur in each year.


In [3]:
#print (years)
years = [row[1] for row in data ]


year_counts = {}
for year in years:
    if year in year_counts:
        year_counts[year] += 1 
    else:
        year_counts[year] = 1 
        
        
print(year_counts)

{'2012': 33563, '2013': 33636, '2014': 33599}


## E) Exploring Gun Deaths By Month And Year
It looks like gun deaths didn't change much by year from 2012 to 2014.  
Let's see if gun deaths in the US change by month and year.  
In order to do this, we'll have to create a datetime.datetime object using the year and month columns.  
We'll then be able to count up gun deaths by date, like we did by year in the last screen.

As you may recall from an earlier mission, you can create a datetime object by specifying the year,  
month, and day keyword arguments:

 + date = datetime(year=2016, month=12, day=1)
 
We can use the month and year column of data to create a datetime. We'll specify a fixed day because we're missing that column in our data. If we create a datetime.datetime object for each row, we can then count up how many gun deaths occurred in each month and year using a similar procedure to what we did in the last screen.

## Analysis by Month and Date 
Use a list comprehension to create a datetime.datetime object for each row. Assign the result to dates.

+ The year column is in the second element in each row.
+ The month column is the third element in each row.
  Make sure to convert year and month to integers using int().
+ Pass year, month, and day=1 into the datetime.datetime() function.
  Display the first 5 rows in dates to verify everything worked.
  


In [4]:
import datetime
dates =[datetime.datetime(year=int(row[1]),month=int(row[2]),day=1)for row in data]
print(dates[0:5])

[datetime.datetime(2012, 1, 1, 0, 0), datetime.datetime(2012, 1, 1, 0, 0), datetime.datetime(2012, 1, 1, 0, 0), datetime.datetime(2012, 2, 1, 0, 0), datetime.datetime(2012, 2, 1, 0, 0)]


## Continuation on Datetime 
Count up how many times each unique date occurs in dates. Assign the result to date_counts.
This follows a similar procedure to what we did in the last screen with year_counts.
Display date_counts.

In [5]:
date_counts = 0
unique_dates = {}
for date in dates:
    if date not in unique_dates:
        unique_dates[date.year] = date.month
        date_counts += 1 
print (unique_dates,date_counts)

{2012: 12, 2013: 12, 2014: 10} 100798


## F) Exploring Gun Deaths By Race And Sex

The sex and race columns contain potentially interesting information on how gun deaths in the US vary by gender and race. 
Exploring both of these columns can be done with a similar dictionary counting technique to what we did earlier.

+ Count up how many times each item in the sex column occurs.Assign the result to sex_counts.
+ Count up how many times each item in the race column occurs.Assign the result to race_counts.
+ Display race_counts and sex_counts to verify your work, and see if you can spot any patterns.



In [6]:
sex_counts = {}
for row in data:
    if row[5] in sex_counts:
        #print(True)
        sex_counts[row[5]] += 1
    else:
        sex_counts[row[5]] = 1
        
print(sex_counts)    

{'M': 86349, 'F': 14449}


In [7]:
race_counts = {}
for row in data:
    if row[7] in race_counts:
        #print(True)
        race_counts[row[7]] += 1
    else:
        race_counts[row[7]] = 1
        
print(race_counts)    

{'Asian/Pacific Islander': 1326, 'White': 66237, 'Native American/Native Alaskan': 917, 'Black': 23296, 'Hispanic': 9022}


## G) Reading In A Second Dataset

We explored gun deaths by race in the past screen. However, our analysis only gives us the total number of gun deaths by race in the US. Unless we know the proportion of each race in the US, we won't be able to meaningfully compare those numbers.  

What we really want to get is a rate of gun deaths per 100000 people of each race.  In order to do this, we'll need to read 
in data about what percentage of the US population falls into each racial category.  Luckily, we can import some census
data to help us out.  

The data contains information on the total population of the US, as well as the total population of each racial group in
the US. The data is stored in the census.csv file, and only consists of two rows:

As you can see, the first row is a header row,  
and the second row consists of population counts.  

We'll need to read this file in using the csv.reader() function.

In [8]:
import pandas as pd

f2 = open('census.csv',"r")
csv_reader = csv.reader(f2)
census = list(csv_reader)

df = pd.read_csv('census.csv')
df

Id                  Year    Id.1         Sex     Id.2  \
0  cen42010  April 1, 2010 Census  totsex  Both Sexes  tothisp   

  Hispanic Origin       Id.3  Id2      Geography      Total  \
0           Total  0100000US  NaN  United States  308745538   

   Race Alone - White  Race Alone - Hispanic  \
0           197318956               44618105   

   Race Alone - Black or African American  \
0                                40250635   

   Race Alone - American Indian and Alaska Native  Race Alone - Asian  \
0                                         3739506            15159516   

   Race Alone - Native Hawaiian and Other Pacific Islander  Two or More Races  
0                                             674625                  6984195

In [9]:
#Map race-name and number of suicides 
mapping  = {}
census= census[1:]

for i in census:
    mapping['Asian/Pacific Islander'] = i[14]+i[13]
    mapping['Black']= i[12]
    mapping['Hispanic']= i[11]
    mapping['Native American/Native Alaskan'] = i[13]
    mapping['White'] = i[10]    
print(mapping)
    



{'Asian/Pacific Islander': '151595163739506', 'Black': '40250635', 'Hispanic': '44618105', 'Native American/Native Alaskan': '3739506', 'White': '197318956'}


## H) Computing Rates Of Gun Deaths Per Race

In order to get from the raw counts of gun deaths by race to a rate of gun deaths per 100000 people in each race,  
we'll need to divide the total number of gun deaths by the population of each race.  
From the census dataset, we know that the number of people in the White racial category is 197318956.  
We'd divide 66237 by 197318956:

This gives us the percentage chance that a given person in the White census race category would have been killed by a gun in the US from 2012 to 2014. If you do this computation, you'll see that the rate is a very small number, 0.0003356849303419181.  
It's for this reason that it's typical to express crime statistics as the "rate per 100000".  
This tells you the number of people in a given group out of every 100000 that were killed by guns in the US.  
To get this, we just multiply by 100000:
  
This gives us 33.56, which we can interpret as "33.56 out of every 100000 people in the White census race category in the US  were killed by guns between 2012 and 2014".

We'll need to calculate these same rates for each racial category. The only stumbling block is that the racial categories are named slightly differently in census and in data.  
We'll need to manually construct a dictionary that allows us to map between them, and perform the division.

Here's a list of the race name in data, and the corresponding race name in census:

+ Asian/Pacific Islander -- Race Alone - Asian plus Race Alone - Native Hawaiian and Other Pacific Islander.
+ Black -- Race Alone - Black or African American.
+ Hispanic -- Race Alone - Hispanic
+ Native American/Native Alaskan -- Race Alone - American Indian and Alaska Native
+ White -- Race Alone - White  

We'll need to create a dictionary that has each race name from data as a key, and has the population count  
for the races from census as the values.

Instructions

+ Manually create a dictionary, mapping that maps each key from race_counts to the population count of the race from census.
  The keys in the dictionary should be Asian/Pacific Islander, Black, Native American/Native Alaskan, Hispanic, and White.
  In the case of Asian/Pacific Islander, you'll need to add the counts from census for Race Alone - Asian, and Race Alone -       Native Hawaiian and Other Pacific Islander.
  
+ Create an empty dictionary, race_per_hundredk.
  Loop through each key in race_counts.
  Divide the value associated with the key in race_counts by the value associated with the key in mapping.
  Multiply by 100000.  
  Assign the result to the same key in race_per_hundredk.
  
When you're done, race_per_hundredk should contain the rate of gun deaths per 100000 people for each racial category.
Print race_per_hundredk to verify your work.

In [10]:
race_per_hundredk = {}

for key in race_counts:
    death_by_race = race_counts[key]
    
    for key2 in mapping:
        total_race = int(mapping[key2])
        
        if key == key2:
            race_per_hundredk[key] = (death_by_race/total_race)*100000

                                                          
print (race_per_hundredk)     
    

{'Asian/Pacific Islander': 8.746980888378049e-07, 'White': 33.56849303419181, 'Native American/Native Alaskan': 24.521955573811088, 'Black': 57.8773477735196, 'Hispanic': 20.220491210910907}


 ## I) Filtering By Intent
 We can filter our results, and restrict them to the Homicide intent.  
 This will tell us what the gun-related murder rate per 100000 people in each racial category is.  
 In order to do this, we'll need to redo our work in generating race_counts,but only count rows where the intent was Homicide.

 We can do this by first extracting the intent column, then using the enumerate() function to loop through each index and value  in the race column. If the value in the same position in intents is Homicide, we'll count the value in the race column.

 Finally, we'll use the mapping dictionary to convert from raw counts to rates.
 
 + Extract the intent column using a list comprehension. The intent column is the fourth column in data.
   Assign the result to intents.
 + Extract the race column using a list comprehension. The race column is the eighth column in data.
   Assign the result to races.
 + Create an empty dictionary called homicide_race_counts.  
 + Use the enumerate() function to loop through each item in races.  
   The position should be assigned to the loop variable i, and the value to the loop variable race.
 + Check the value at position i in intents.
   If the value at position i in intents is Homicide:
   If the key race doesn't exist in homicide_race_counts, create it.
   Add 1 to the value associated with race in homicide_race_counts.
   
When you're done, homicide_race_counts should have one key for each of the racial categories in data.  
The associated value should be the number of gun deaths by homicide for that race.



In [11]:
intents = [row[3] for row in data]
#print(intents)

races = [row[7] for row in data]
#print(races)

homicide_race_counts  = {}

for i,race in enumerate(races):    
        intent = intents[i]
        if  intent == "Homicide" :
            if race in homicide_race_counts:
                homicide_race_counts[race] += 1 
            else: 
                homicide_race_counts[race] = 1
                
print(homicide_race_counts)


{'White': 9147, 'Asian/Pacific Islander': 559, 'Black': 19510, 'Native American/Native Alaskan': 326, 'Hispanic': 5634}


## Homicide per 10,000 in the race 

+ Perform the same procedure we did in the last screen using mapping on homicide_race_counts to get 
  from raw numbers to rates per 100000.
+ Display homicide_race_counts to verify your work.
+ Write up your findings in a markdown cell.
+ Write up any next steps you want to pursue with the data in a markdown cell.

In [12]:
homicide_per_race_10000 = {}

for i in homicide_race_counts:
    for j in mapping: 
        if i == j :
            homicide_per_race = homicide_race_counts[i]
            total_Race_Census = int(mapping[j])            
            homicide_per_race_10000[i] = (homicide_per_race/total_Race_Census) * 10000
            
print(homicide_per_race_10000)
            

{'White': 0.46356417981453335, 'Asian/Pacific Islander': 3.6874527274534915e-08, 'Black': 4.847128498718095, 'Native American/Native Alaskan': 0.8717729026240364, 'Hispanic': 1.2627161104219913}


## NOTES: You can do the same for other intents 